### DISCUSSION TITLE
W32_ETH01

### DISCUSSION THREAD

**Discussion by salrodri** [luc-github/ESP3D#827] [F.A.Q] [answered]
Would it be possible to add the libraries for this board W32_ETH01.

So, I can use this board to connect the printer trough Ethernet?

Would be possible to have both options? Ethernet or Wi-Fi? 

thank you,

**Comment by salrodri** [comment]
Never mind,
I just realized that I could enable Ethernet connection in configuration.h 

thank you

**Comment by rubin110** [comment]
Did you ever get this working with that board? I'd like to try out this project but I really would like to avoid using wifi.

**Comment by salrodri** [comment]
technically yes,
I was able to flash the firmware with no problems.

I haven't tested the ethernet connection...
I read on the datasheet that wifi takes priority over ethernet.
so you need to force the chip to use ethernet

My setup includes both. I wifi loses connection, ethernet kicks in.

you can force the connection through ethernet if you disable the wifi under
the socket section.

I will test the ethernet connection only  and I will let you know if woks

**Reply by rubin110** [reply]
Thank you!

**Reply by luc-github** [reply]
to be honest I did not tested ethernet for a while, so feedback would be appreciate

**Comment by salrodri** [comment]
I am sad to report that ethernet did not work for me.

I disable wifi and enable ethernet under webSocket.h

I build firmware and flash the W32 with no problems.

So I powered up with the cable connected to my router and everything looks
good. The printer works just fine. So in order to test the ethernet
connection I unplug the ethernet cable from my router (Technically if I
disable wifi and only enable ethernet. unplugging the cable should stop the
service. Well that did not happen. Unplugging the cable did not have any
effect.)

My next experiment would be "Hello World"  using the W32 library. ( I just
want to confirm that the connection through ethernet is working. If I have
success. then I can come back to the code and see if I am missing something)

thank you,

**Reply by luc-github** [reply]
I do not understand, in esp3d you do not have to change anything in code for ethernet, just enable it in configuration.h

can you clarify the changes you are doing and the tests you are doing?
what service are you refering to?

**Comment by salrodri** [comment]
It seems that on previous versions of esp3d  that is the case. You can
enable that feature under configuration.h
In the most recent version I did not find this feature under
configuration.h.

**Reply by luc-github** [reply]
yes you can https://github.com/luc-github/ESP3D/blob/3.0/esp3d/configuration.h#L108
But my tests were limited to arduino ide and olimex POE
Your board / ethernet may need additional settings in PlatformIO or even in esp3d if is not same as olimex (spi pins / address / etc) - because I do not have so many devices I did not added extra configuration items yet
what board are you using ?

**Reply by luc-github** [reply]
Just did a test on latest git :
![image](https://user-images.githubusercontent.com/8822552/200857802-6d79acb4-4cfc-434b-99a1-28177d4042ed.png)
![image](https://user-images.githubusercontent.com/8822552/200857935-847eec3f-2305-4637-996e-8fb8af4e155d.png)
I see a small bug on cable status - I will fix that tomorrow and I will take opportunity to enable the olimex in platformio

I see the differences of olimex POE compared to esp32 variant.h are :
```
#define ETH_CLK_MODE ETH_CLOCK_GPIO17_OUT
#define ETH_PHY_POWER 12

static const uint8_t SS    = 5;
static const uint8_t MOSI  = 2;
static const uint8_t MISO  = 15;
static const uint8_t SCK   = 14;
```

So I will add these parameters configurables tomorrow 

Currently even ssdp is working on my side:
![image](https://user-images.githubusercontent.com/8822552/200859498-77e9e502-81d6-489b-b84d-38543349ba3c.png)

**Comment by salrodri** [comment]
I am using the chip W32_ETH01.

Do I have to download this configuration.h separately ?
I can't find this configuration.h.  This is not included in the zip folder
that I just downloaded.

**Reply by luc-github** [reply]
download latest git, not pre-release - as you can see the entry is present in file

**Comment by salrodri** [comment]
nevermind.
I was looking at the wrong folder.
I found it.

I will flash it and test it after work.

Thank you very much for your help!!!

**Reply by luc-github** [reply]
give also a try to https://github.com/espressif/arduino-esp32/tree/master/libraries/Ethernet/examples
may be your board need some changes - if you can identify them I can support it

**Comment by salrodri** [comment]
Understood,
thank you for pointing me in the right direction.

**Reply by luc-github** [reply]
I only have one device so support of ethernet is definitly not optimal...hope it will getting better with feedback like yours

**Comment by salrodri** [comment]
I will run a couple of tests today and I will get back to you.

Thank you,

**Comment by rubin110** [comment]
@luc-github Not sure where in the world you are, but I'm happy to order you one or a couple boards with ethernet on it if this is something you would like to spend time on to support.

**Comment by salrodri** [comment]
I am in the US in Denver Colorado.
Do you think this is worth pursuing?
As far as I know the only benefit with ethernet is a stable connection.
I would be curious to see if the uploading time improves.

Are you guys trying to compete with octoprint?

**Reply by luc-github** [reply]
@rubin110 I am in Taiwan, yes if there is need I will to support it of course

@salrodri the benefit of ethernet is also to remove wifi latency and so to be able to stream gcode without impact on pront job
and no, cannot compete with octoprint a 1$ chip cannot has same performances /power of mini pc

**Reply by luc-github** [reply]
Ok About cable disconnected it is core issue https://github.com/espressif/arduino-esp32/issues/6105
I will see how to workaround
Edit: Issue solved but when doing soft reset I drop to new issue : https://github.com/espressif/arduino-esp32/issues/1912#issuecomment-426247971 orz...

**Reply by luc-github** [reply]
Ok fixed all issues (I hope, but with Murphy law there is always a chance of surprise...) 
```
M117 Cable connected
M117 192.168.2.144
M117 192.168.2.144
[ESP420]
;chip id: 34003
;CPU Freq: 240Mhz
;CPU Temp: 53.3C
;free mem: 223.07 KB
;SDK: v4.4.1-472-gc9140caf8c
;flash size: 4.00 MB
;size for update: 1.87 MB
;FS type: LittleFS
;FS usage: 8.00 KB/192.00 KB
;baud: 115200
;wifi: OFF
;ethernet: ON
;hostname: esp3d
;HTTP port: 80
;Telnet port: 23
;ethernet: 24:0A:C4:26:D3:87
;cable: connected (100Mbps)
;ip mode: dhcp
;ip: 192.168.2.144
;gw: 192.168.2.1
;msk: 255.255.255.0
;DNS: 192.168.2.1
;serial: ON
;M117: ON
;notification: ON(none)
;Target Fw: unknown
;FW ver: 3.0.0.a219
;FW arch: ESP32
```

Now Configuration.s has more settings for ethernet, I have just tested with my olimex POE, so need more testing
```
//#define ETH_FEATURE

//Ethernet type (Check ETH.h eth_phy_type_t)
//TYPE_ETH_PHY_LAN8720
//TYPE_ETH_PHY_TLK110
//TYPE_ETH_PHY_RTL8201
//TYPE_ETH_PHY_DP83848
//TYPE_ETH_PHY_DM9051
//TYPE_ETH_PHY_KSZ8041
//TYPE_ETH_PHY_KSZ8081
#define ESP3D_ETH_PHY_TYPE TYPE_ETH_PHY_LAN8720

//Ethernet board Clock mode
// MODE_ETH_CLOCK_GPIO0_IN
// MODE_ETH_CLOCK_GPIO0_OUT
// MODE_ETH_CLOCK_GPIO16_OUT
// MODE_ETH_CLOCK_GPIO17_OUT
#define ESP3D_ETH_CLK_MODE MODE_ETH_CLOCK_GPIO17_OUT

//Pins of ethernet board
#define ESP3D_ETH_PHY_POWER_PIN 12
//#define ESP3D_ETH_PHY_MDC_PIN 23
//#define ESP3D_ETH_PHY_MDIO_PIN 18

//Address of ethernet board
//#define ESP3D_ETH_PHY_ADDR 0
```
I will update the ESP3D configurator once more tests / feedbacks are done 

let me know what is on your side,  for what I can read from : https://github.com/khoih-prog/WebServer_WT32_ETH01/blob/main/src/WebServer_WT32_ETH01.hpp#L39-L74
the settings for WT32-ETH01 should be :
```
  #define ESP3D_ETH_PHY_TYPE TYPE_ETH_PHY_LAN8720
  #define ESP3D_ETH_CLK_MODE  MODE_ ETH_CLOCK_GPIO0_IN  
  #define ESP3D_ETH_PHY_POWER_PIN 16
  #define ESP3D_ETH_PHY_MDC_PIN 23
  #define ESP3D_ETH_PHY_MDIO_PIN 18
  #define ESP3D_ETH_PHY_ADDR 1
```

**Comment by salrodri** [comment]
Please email me this version to test it.
I flash the firmware with version 3. But nothing happens.
No connection through wifi or ethernet. (for some reason my serial monitor
was not working. it seems like the baud rate settings did not take any
effect )

I would like to give a try to this version that you have been working on.

I will be doing more testing after work.

thank you,

**Comment by salrodri** [comment]
after not having any luck.
I went back to a previous version 2.1X and it seems that this version works
just fine.
I did not have a chance to test the ethernet.

**Reply by luc-github** [reply]
esp3d v2.1 does not support ethernet

**Comment by salrodri** [comment]
got it.
well. I will fix my serial monitor, so I can troubleshoot this issue.

**Reply by luc-github** [reply]
After Flashing ESP3D V3 to ESP3D 2.1 you must do a `[ESP444]RESET` 
another solution is to erase flash before flashing, there is a menu in vscode / platformio for that

**Comment by salrodri** [comment]
I will look into erase the flash before flashing

I thought that the reset was done automatically by Vs code.
I don't remember exactly the message, but is something like "leaving bord,
doing hard reset "

How do I do a [ESP444]RESET?

**Reply by luc-github** [reply]
using platformio serial terminal in monitoring mode

**Comment by salrodri** [comment]
Got it.
going back to your offer, what other boards do you have in mind?

Do you know any chip that utilizes 802.11 AX?

is it true that the slow uploading process is a limit on marlin and not in
esp3d?

I keep going back to this issue. it does not matter how fast or how stable
the connection is.
if the cap is on marlin. it does not makes a lot of sense improvings this
aspects/

what do you think?

**Reply by luc-github** [reply]
I do not understand your question
the upload speed is depending on where you want to upload, and how esp is connected to sd card
if you refere to serial upload, MKS serial protocol allows up to 110KB/s vs a 0.4 KB/s by M28M29
a direct access to SD can even more than 500KB/s on arduino and 1.2MB/s on esp32 and Duet Wifi

there are lot of parameters to take in consideration

**Comment by salrodri** [comment]
The main issue is uploading the G-code to the SD card.
I've been using esp32 .

It sounds like I can change those settings in order to make this process
faster.

**Comment by salrodri** [comment]
Luc,
Ethernet still not working for me.

I Went back and forth between wifi and ethernet and wifi works just fine,

Ehtenert does not.

My next step is to test the board and make sure that the board is working
properly.

thank you,

**Comment by rubin110** [comment]
My need is to not use a RaspberryPi anymore for Octoprint. Run Octoprint on a faster home server instead. But the home serveris elsewhere in the house far from the printer. I generally don't trust wifi for things that really depend on low latency. 

@luc-github I'm trying to understand, so with ESP3D it's just transmitting serial over IP (with some extra web GUI), correct? If there's no bottleneck compared to a direct USB between the OctoPrint machine and the printer (streaming gcode and not copying to SD card), I'd happily run ESP3D.

I don't think I can safely ship anything to Taiwan right now, but I can send you some Bitcoins for your help with supporting ethernet.

**Reply by luc-github** [reply]
ESP3D is a control / monitor directly connected to printer, it also allow to bridge TCP to serial , and in theory streaming gcode is possible, when I do not recommend to do it using wifi due to latency, in theory using ethernet should be ok, but still less safe than direct usb because still subject to router behavior.
But to be honest I never experienced it - I have supported ethernet because was interresting feature but I never really used it due to lack of devices available none on esp8266 and few on esp32
I have ordered a couple of wt32_eth01 and I should be able to get them in couple of days TBC

Unfortunatly I do not have any Bitcoins wallet only paypal sorry

**Reply by luc-github** [reply]
for what I can see the board is supported in latest git https://github.com/espressif/arduino-esp32/blob/master/variants/wt32-eth01/pins_arduino.h

**Comment by luc-github** [comment, marked as answer]
Got my boards today :
I used the same settings as variant:
```
#define ETH_FEATURE

//Ethernet type (Check ETH.h eth_phy_type_t)
//TYPE_ETH_PHY_LAN8720
//TYPE_ETH_PHY_TLK110
//TYPE_ETH_PHY_RTL8201
//TYPE_ETH_PHY_DP83848
//TYPE_ETH_PHY_DM9051
//TYPE_ETH_PHY_KSZ8041
//TYPE_ETH_PHY_KSZ8081
#define ESP3D_ETH_PHY_TYPE TYPE_ETH_PHY_LAN8720

//Ethernet board Clock mode
// MODE_ETH_CLOCK_GPIO0_IN
// MODE_ETH_CLOCK_GPIO0_OUT
// MODE_ETH_CLOCK_GPIO16_OUT
// MODE_ETH_CLOCK_GPIO17_OUT
#define ESP3D_ETH_CLK_MODE MODE_ETH_CLOCK_GPIO0_IN

//Pins of ethernet board
#define ESP3D_ETH_PHY_POWER_PIN 16
#define ESP3D_ETH_PHY_MDC_PIN 23
#define ESP3D_ETH_PHY_MDIO_PIN 18

//Address of ethernet board
#define ESP3D_ETH_PHY_ADDR 1
```

and also need to disable the pin reset feature, or find a not conflicting pin, because gpio 0 is not good for this board, Pin Reset is triggered when enabling Ethernet board, and so reset the EEPROM, so it restart always in AP mode.
```
/************************************
*
* Reset ESP3D
*
* Reset ESP3D settings
*
************************************/

/* Enable pin reset feature
* Use a pin to reset ESP3D settings
*/
//#define PIN_RESET_FEATURE

/* Reset pin
* The pin used to reset ESP3D setting if set to low for more than 1 second at start
*/
//#define ESP3D_RESET_PIN 0
```

And it works ^_^ so far so good

![image](https://user-images.githubusercontent.com/8822552/201475669-1755fa3f-0344-48ad-ba4a-24c6851133aa.png)

**Reply by luc-github** [reply]
Ethernet configuration is now in configurator
![image](https://user-images.githubusercontent.com/8822552/201502602-83d65c91-d0d6-40c5-93d2-c7288fa2b1f0.png)

I have also added some fix in esp3d  to limit conflict between Pin reset feature and pin used by others devices, and also unwhished usage after boot, it does not cover all cases (eg: if reset pin is used by screen, or any early init device)  but it works for ethernet board at least

**Comment by salrodri** [comment]
Luc,
that looks marvelous.
I will give it a try today after work.

thank you,

**Comment by salrodri** [comment]
Luc, what board are you using?

 I am using the chip W32_ETH01.
V2.1X works just fine (Although ethernet is not available in this version)
I can connect through wifi.

When I flash the chip V3.0 nothing happens.
no connection through ethernet or wifi.

In V3 Is the AP_MODE activated by default?

thank you,

**Reply by luc-github** [reply]
![image](https://user-images.githubusercontent.com/8822552/202223480-4dc1463f-7646-4920-a0e7-5ea130565b2b.png)
https://github.com/luc-github/ESP3D/discussions/827#discussioncomment-4107720

**Comment by salrodri** [comment]
that is exactly the same that I am using.
I am not sure what I am doing wrong.
I will give it a try again.
thank you,

**Comment by salrodri** [comment]
Luc,
I did all the modifications to the configuration.h file

#define ETH_FEATURE

//Ethernet type (Check ETH.h eth_phy_type_t)
//TYPE_ETH_PHY_LAN8720
//TYPE_ETH_PHY_TLK110
//TYPE_ETH_PHY_RTL8201
//TYPE_ETH_PHY_DP83848
//TYPE_ETH_PHY_DM9051
//TYPE_ETH_PHY_KSZ8041
//TYPE_ETH_PHY_KSZ8081
#define ESP3D_ETH_PHY_TYPE TYPE_ETH_PHY_LAN8720

//Ethernet board Clock mode
// MODE_ETH_CLOCK_GPIO0_IN
// MODE_ETH_CLOCK_GPIO0_OUT
// MODE_ETH_CLOCK_GPIO16_OUT
// MODE_ETH_CLOCK_GPIO17_OUT
#define ESP3D_ETH_CLK_MODE MODE_ETH_CLOCK_GPIO0_IN

//Pins of ethernet board
#define ESP3D_ETH_PHY_POWER_PIN 16
#define ESP3D_ETH_PHY_MDC_PIN 23
#define ESP3D_ETH_PHY_MDIO_PIN 18

//Address of ethernet board
#define ESP3D_ETH_PHY_ADDR 1

and also need to disable the pin reset feature or find a not conflicting
pin as 0 is not good for this board, it is trigered when enabling Ethernet
board and so reset the EEPROM so it always in AP mode

/************************************
*
* Reset ESP3D
*
* Reset ESP3D settings
*
************************************/

/* Enable pin reset feature
* Use a pin to reset ESP3D settings
*/
//#define PIN_RESET_FEATURE

/* Reset pin
* The pin used to reset ESP3D setting if set to low for more than 1
second at start
*/
//#define ESP3D_RESET_PIN 0

but still does not work the ethernet
[image: image.png]

**Reply by luc-github** [reply]
cannot see the screenshot
did you reset ESP as I wrote above after flashin V3 ?

**Comment by salrodri** [comment]
I did.
Would you please show me your wirering?
I think I am not wirerinig correctly.

**Reply by luc-github** [reply]
What wiring ? I just connect tx0 / rx0 / gnd and 3.3v, remove the GPIO IO0 before reboot and enable serial monitoring
after reset default mode is AP Config
What give [ESP420] as output ?

also what is your error ?

**Comment by salrodri** [comment]
404 not found preferences.json

**Reply by luc-github** [reply]
so ? that is normal - it does not block anything 
it will disapear as soon as you save settings in interfaces 
You said nothing work - so this is not what happened - is that just this message that bother you ?

**Comment by salrodri** [comment]
I thought that I was missing a file.
There is actually a file under the folder call server named like that.
preferences.json.
I included this with my index.html.gz. but it did not have any effect.

It seems to be working fine.
I haven't connected to the printer yet.

I am still struggling with the ethernet.

**Comment by luc-github** [comment]
no do not copy preferences.json from any place - it may not synced with your version and generate weird behavior in UI
What give [ESP420] as output ?

**Comment by salrodri** [comment]
What do you mean?
here is what I am doing
once I change the configurations on the WEBUI from access point to
ethernet, then save the changes.
disconnect the power, connect ethernet cable, power back on.
I should be able to open the WEBUI with the same ip address. correct?

**Reply by luc-github** [reply]
not same ip if use dhcp , it is not same device
what is your os ?

**Comment by salrodri** [comment]
windows 11

**Reply by luc-github** [reply]
so just go to network panel of your file explorer you should see your esp device, click on icon it will open web browser

**Comment by salrodri** [comment]
🤓🤓🤓🤓🤓🤓🤓🤓🤓🤓🤓🤓🤓
it works!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by salrodri** [comment]
That's amazing, I took a personal day today to work on this thing. haha
now I can go back to work!!

Is it the same for bluetooth?
enable it, then connect it and the WEBUI should pop up?

Thank you!!!!!!!!!!!!!!!!

**Reply by luc-github** [reply]
there is no webui in bt, it is bt, not web

**Comment by salrodri** [comment]
So how do you control it without WEBUI?

**Reply by luc-github** [reply]
you need to use an application that use bt on your phone/tablet

**Reply by luc-github** [reply]
like https://play.google.com/store/apps/details?id=com.fox3d.controller

**Comment by salrodri** [comment]
I see.
Would you mind if I create a tutorial on how to set up the ethernet?

thank you,

**Reply by luc-github** [reply]
no problem, I am not good at doc, good luck ^_^
Did you tried : https://luc-github.github.io ?

**Comment by salrodri** [comment]
I did not.
I was wondering where you got that screenshot from.
So, how does it work?
Do I select all my configurations? then download the webui?

**Reply by luc-github** [reply]
Try it - you will see :)

it generate platformio.ini and configuration.h according your definition so you do not need to dig in files, replace the existing files by the generated one and compile/flash
The link to download webui is an helper to select right webui according your configuration as now I splitted webui per target and FW to decrease file size.

**Comment by salrodri** [comment]
Awesome, thanks!

**Comment by salrodri** [comment]
Do you think it is possible to add esp3d to the Cura slicer marketplace?
Cura has the option, after you slice your model. you can select print with
ocotrprint.
Would it be possible to accomplish this?

By the way. I wanted to say thank you for your hard work!!

thank you,

**Reply by luc-github** [reply]
I do not know , I do not know what this option do
